### ISSUE TITLE
vswitch: multiple set-workspace-request end up on wrong workspace

### ISSUE THREAD

**Issue by dkondor** [WayfireWM/wayfire#676]
I've found that in the vswitch plugin, if it receives a set-workspace-request while it is active, the result is inconsistent (it ends up of a different workspace than intended).

This is an edge case, I've found it while playing with cairo-dock (https://github.com/dkondor/cairo-dock-core/tree/foreign_toplevel).

I've describing here a simpler way to reproduce, but this method requires a touchscreen or very fast reflexes :)

Steps to reproduce:
1. Start wayfire. I'm assuming the following layout for workspaces, but the issue is present with a 3x3 grid as well:
```
A B
C D
```
2. Start at least two apps on two different workspaces. E.g. on workspace B and C in the above configuration.
3. Start wf-dock.
4. Switch back to workspace A.
5. Activate the two apps from wf-dock in quick succession (so that both are selected before the workspace switch can happen). I have a touchscreen and found that I'm able to do this by tapping quickly with different fingers on the two icons. I've managed to do this very quickly with the mouse only a few times.

Expected results: wayfire ends up on the workspace with the app that was activated last (either B or C)

Actual results: wayfire switches to workspace D. Also, the animation is weird.

Note: what seems to be happening is that two relative motions are carried out (right and down) which add up to ending up on workspace D, instead of only one of the motions that would go to workspace B or C.

Also, this does not happen with the workspace navigation keyboard shortcuts.

If unclear, I can create a dummy "plugin" that reproduces this issue.

Also, I will create a pull request with a proposed solution.

**Comment by ammen99** [comment]
Wait, cairo-dock works on wayfire now ?!?!? Sign me up :)

**Comment by ammen99** [comment]
On the actual issue: I can see what it is, the `add_direction` we call is with the difference between old and new viewport, it should be the difference between current target workspace and new viewport.

**Comment by damianatorrpm** [comment]
Instead of fast fingers you can use the dbus plugin 
https://github.com/damianatorrpm/wayfire-plugin_dbus_interface
to reproduce this

**Comment by ammen99** [comment]
> Instead of fast fingers you can use the dbus plugin
> https://github.com/damianatorrpm/wayfire-plugin_dbus_interface
> to reproduce this

No need to, I know what the bug is and how it should be fixed.

**Comment by ammen99** [comment]
@dkondor Does cairo-dock really work on wayland though?

**Comment by dkondor** [comment]
I've added a PR that fixes this for me, but I'm not sure if that is the right way.

I'll check out the dbus plugin nevertheless, it would be great to have that.

Regarding cairo-dock, it already had wayland support for rendering, but needs something like the layer-shell and foreign-toplevel protocols to make it usable. I've started experimenting in the two branches I have:
https://github.com/dkondor/cairo-dock-core/tree/foreign_toplevel
https://github.com/dkondor/cairo-dock-core/tree/layer_shell

You need to merge both branches to have something usable. The main functionality works, but there are some issues, including focus, autohide and workspaces (these probably need additional protocols). Although it does not seem impossible to fix them. Any feedback is welcome :) I'd really like to keep using cairo-dock, even on wayland.

**Comment by ammen99** [comment]
> I've added a PR that fixes this for me, but I'm not sure if that is the right way.
> 
> I'll check out the dbus plugin nevertheless, it would be great to have that.
> 
> Regarding cairo-dock, it already had wayland support for rendering, but needs something like the layer-shell and foreign-toplevel protocols to make it usable. I've started experimenting in the two branches I have:
> https://github.com/dkondor/cairo-dock-core/tree/foreign_toplevel
> https://github.com/dkondor/cairo-dock-core/tree/layer_shell
> 
> You need to merge both branches to have something usable. The main functionality works, but there are some issues, including focus, autohide and workspaces (these probably need additional protocols). Although it does not seem impossible to fix them. Any feedback is welcome :) I'd really like to keep using cairo-dock, even on wayland.

I tried to merge these two branches, and enabled gtk-layer-shell, but cairo-dock still doesn't start in the correct mode. Do I need to set env vars?

**Comment by dkondor** [comment]
You have to unset DISPLAY and maybe also set GDK_BACKEND to wayland.

**Comment by dkondor** [comment]
Also, I realized for cairo-dock, it is important to enable EGL and layer-shell with cmake:
```
cmake .. -Denable-egl-support=True -Denable-gtk-layer-shell=True
```
It should result in:
```
--  * With GLX support    : no
--  * With Wayland support: yes (1.18.90)
--  * With gtk-layer-shell: yes
--  * With EGL support    : yes (20.0.8)
```
(GLX support has to be disabled; if it is enabled, it is best to delete CMakeCache.txt or the while build directory and try again)

**Comment by ammen99** [comment]
@dkondor, thanks, finally managed to get it running. Do you need bug reports atm, and if yes, where?

**Comment by dkondor** [comment]
Great, happy to hear that! I'm happy with any feedback, but it's not urgent, since I only have time to spend on this sporadically. That's why I'm also unsure if I should add an issues page in my repository.
Maybe you can comment on my PRs for now?
https://github.com/Cairo-Dock/cairo-dock-core/pull/8
https://github.com/Cairo-Dock/cairo-dock-core/pull/9

**Comment by soreau** [comment]
> I've found that in the vswitch plugin, if it receives a set-workspace-request while it is active, the result is inconsistent (it ends up of a different workspace than intended).
> 
> This is an edge case, I've found it while playing with cairo-dock (https://github.com/dkondor/cairo-dock-core/tree/foreign_toplevel).

🎉 Nice work! 🥳

> 
> I've describing here a simpler way to reproduce, but this method requires a touchscreen or very fast reflexes :)
> 
> Steps to reproduce:
> 
>     1. Start wayfire. I'm assuming the following layout for workspaces, but the issue is present with a 3x3 grid as well:
> 
> 
> ```
> A B
> C D
> ```
> 
>     1. Start at least two apps on two different workspaces. E.g. on workspace B and C in the above configuration.
> 
>     2. Start wf-dock.
> 
>     3. Switch back to workspace A.
> 
>     4. Activate the two apps from wf-dock in quick succession (so that both are selected before the workspace switch can happen). I have a touchscreen and found that I'm able to do this by tapping quickly with different fingers on the two icons. I've managed to do this very quickly with the mouse only a few times.
> 
> 
> Expected results: wayfire ends up on the workspace with the app that was activated last (either B or C)
> 
> Actual results: wayfire switches to workspace D. Also, the animation is weird.
> 
> Note: what seems to be happening is that two relative motions are carried out (right and down) which add up to ending up on workspace D, instead of only one of the motions that would go to workspace B or C.

Yes, I think you are correct. Reading the code, I believe the following happens:
1) request_workspace(1, 0) is called (set_workspace() isn't called until animation completes)
2) vswitch handles this as add_direction(new - old) (old is 0, 0)
3) request_workspace(0, 1) is called (set_workspace() still hasn't been called)
4) vswitch handles this as add_direction(new - old) (old is still 0, 0)
This means add_direction(1, 0) followed by add_direction(0, 1) (instead of add_direction(-1, 1))

TL;DR Does [this patch](http://ix.io/2tA2) help?

> 
> Also, this does not happen with the workspace navigation keyboard shortcuts.

This doesn't happen for keybindings because the activators have add_direction() calls with hard-coded values.

**Comment by ammen99** [comment]
> TL;DR Does [this patch](http://ix.io/2tA2) help?

there is already #677 :)

**Comment by dkondor** [comment]
Hi,
I've just found that this issue can persist in the edge case when the same workspace is requested twice in quick succession. I've been able to see it with the scale plugin with [this](http://ix.io/2vYF) patch by @soreau 

I've added a simple print statement and I think what is happening is that for the second request, vswitch::add_direction() is called with x == y == 0, and thus returns false (line 140). This sets carried_out to false in on_set_workspace_request() (line 204).

In workspace-impl.cpp, line 910, then it will test this and call set_workspace() that instantly sets the new workspace. But the vswitch animation is still running, so it ends up on the wrong workspace in the end.

Maybe, add_direction() could return true if x == y == 0, and thus the operation is a no-op?

**Comment by soreau** [comment]
I debugged this and came to the same conclusion. Have you tested if returning true alone fixes the issue?

**Comment by dkondor** [comment]
Yes, returning true fixes it! (tested with the scale patches)

**Comment by soreau** [comment]
@dkondor Can you make a PR for this? I will gladly do it if you can't for some reason.
